### ISSUE TITLE
Unable to provision TOTP

### ISSUE THREAD

**Issue by patel999jay** [pypi/warehouse#14710]
> Hey -- i seem to have the same issue... I've tested with both authy, google authenticator + Microsoft authenticator
> 
> ![Screenshot from 2023-10-06 13-45-59](https://user-images.githubusercontent.com/15642823/273075546-5f18b782-81b5-4235-aa55-544e403377b3.png)
> 
> This is my user account https://pypi.org/user/avi_vajpeyi/
> 
> Any suggestions?
> 
> (I tried the 'Time sync' option for the google authenticator app, i got a message stating "... in sync with Google servers")

i have the same issue, could not do it with authy, google authenticator and/or Microsoft authenticator.

_Originally posted by @patel999jay in https://github.com/pypi/warehouse/issues/7385#issuecomment-1751389413_

**Comment by miketheman** [comment]
@patel999jay I've double-checked the behavior for TOTP enrollment, and it still works as expected.
Are you certain you're not adding any other characters to the input?
Input is also time-sensitive.

**Comment by patel999jay** [comment]
@miketheman : I am certain I am not adding any other character to the input and also input the code in timely fashion. 

**To Reproduce**

- Create New Account
- Try To Enable 2FA
- Every time getting this : "_Invalid TOTP code. Try again?_"

**Comment by miketheman** [comment]
@patel999jay I am unable to reproduce the issue. I've also just confirmed the behavior is working on both pypi.org and test.pypi.org .

And you have confirmed that your app has been time-synced?

**Comment by patel999jay** [comment]
@miketheman : yes it is sync and i tried using both android and apple device, not working on any of them.

![image](https://github.com/pypi/warehouse/assets/5512610/7451b4b2-92dc-43cd-af93-121c5e7623dd)

**Comment by avivajpeyi** [comment]
Hey -- I'm having the same issue, tested with authy, Microsoft authenticator and the Google authenticator apps. :(

**Comment by di** [comment]
For folks having issues adding TOTP, can you visit https://time.is/ on your **mobile** device (the same one with your authenticator app) and confirm that it says "Your time is exact!"?

**Comment by patel999jay** [comment]
Take a look at this:

![image](https://github.com/pypi/warehouse/assets/5512610/05309bba-54df-49f5-8523-07ff3c0909af)

**Comment by di** [comment]
Thanks @patel999jay. Can you confirm that in Google Authenticator, you're using the code from an entry that says `PyPI: <username>`?

**Comment by patel999jay** [comment]
@di : yes, correct i am using this from that entry.

**Comment by di** [comment]
@patel999jay Can you try deleting all "PyPI" entries from your authenticator app, hard-reloading the page with the QR code, scanning it in via your app and using the new code it provides?

**Comment by di** [comment]
@patel999jay Is this the PyPI account in question? https://pypi.org/user/jpatel1993/. This is showing that you already have 2FA enabled on this account. Are you trying to add a new 2FA device? Do you see any TOTP devices listed at https://pypi.org/manage/account/?

**Comment by patel999jay** [comment]
@di : right, i just tried adding my account via my `Ubuntu` machine with `authy` app and it would let me do it. I still can not add that to my phone.

**Comment by di** [comment]
You can only have one TOTP device at a time. Are you trying to replace the Authy device with a mobile device?

**Comment by patel999jay** [comment]
yes right.

**Comment by avivajpeyi** [comment]
Hmm I havent set up 2FA on _any_ device, and still having the issue. 
My username account:
https://pypi.org/user/avi_vajpeyi/


Any suggestions on how to fix/diagnose this error?

![Screenshot from 2023-10-09 14-51-00](https://github.com/pypi/warehouse/assets/15642823/b7df631d-96d7-4b72-b575-0d53170eb348)


![Screenshot from 2023-10-09 12-53-25](https://github.com/pypi/warehouse/assets/15642823/c5aaddbc-cdd8-44ca-8ad0-0962c5f9e4ce)
![Screenshot from 2023-10-09 12-53-06](https://github.com/pypi/warehouse/assets/15642823/852b16aa-3e5c-41b3-8d63-36995a5010c0)

**Comment by di** [comment]
I can't reproduce this: With a brand new PyPI account, Google Authenticator + Google Chrome, I was able to add a new 2FA device, and to edit/replace an existing 2FA device. I'm going to try to add some additional error messages here to indicate if this is due to clock drift or not, but I'm not sure what other leads to pursue.

**Comment by di** [comment]
Hi folks, I've landed some changes in #14720 that should emit a different error if this is due to clock skew. Could you try again?

**Comment by avivajpeyi** [comment]
Hey @di -- thanks for looking into this. Unfortunately, I'm still just getting the 'Invalid TOTP code. Try again?' error (no error based on clock skew)
<img width="329" alt="Screen Shot 2023-10-10 at 12 21 21 pm" src="https://github.com/pypi/warehouse/assets/15642823/c623ba89-42b0-4c01-9e1e-531ee0d76ae3">

FWIW, I use authy for my uni account 2FA -- that does work.

**Comment by dreamcode1994** [comment]
hey @di 
For me also Same behaviour "Invalid TOTP code Try Again"


My Mobile Time is also Correct. Using Google Authenticator , deleted all entries then added again with hard refresh of page. 
![time](https://github.com/pypi/warehouse/assets/118388087/8d2da19f-2a7d-43f9-981e-51603d4c98b3)
<img width="1727" alt="Screenshot 2023-10-10 at 8 53 26 AM" src="https://github.com/pypi/warehouse/assets/118388087/514049fc-e6a7-4f6d-a1c3-7f67a1681183">

Can you please look into this?

**Comment by di** [comment]
Would someone experiencing this be able to schedule a call with me here so we can walk through it together? https://calendly.com/di_codes

**Comment by di** [comment]
Just sat down with @dreamcode1994 and we determined that it is something about his Google Chrome browser that is causing this, as he was able to get this to work in Safari and in a Google Chrome incognito window. Can other folks in this thread try different browsers as well?

**Comment by dreamcode1994** [comment]
Thanks @di for quick response.

**Comment by avivajpeyi** [comment]
Deleting my previous scanned version of pypi + updating my firefox + phone worked!  Thanks @di

**Comment by patel999jay** [comment]
@di : it worked. Thanks !

**Comment by woodruffw** [comment]
Out of curiosity: did anybody in this thread who experienced trouble take their device across a timezone shortly before experiencing it?

**Comment by di** [comment]
@woodruffw Do you mean the device generating the TOTP code, or the device with the web browser? I think what I determined with @dreamcode1994 was that their mobile device was producing valid codes, but something about their web browser they were submitting the code with was causing it to be invalid.

**Comment by woodruffw** [comment]
> Do you mean the device generating the TOTP code, or the device with the web browser?

Ah yeah, I was thinking of the former. But if it's the browser submitting it incorrectly, then I'm thoroughly mystified 🙂

**Comment by GLGDLY** [comment]
problem solved also by using Firefox instead of Chrome

**Comment by andrwcnln** [comment]
I just added a section to the help in #15079 for TOTP issues. I could update it to cover using a different browser as an additional troubleshooting step?

**Comment by ideoforms** [comment]
Another datapoint: This also happened to me the first time I set up PyPi/TOTP (Chrome on macOS, Google Authenticator on iOS, UK timezone, set up by scanning the QR code). The authentication repeatedly failed.

I eventually resolved it by deleting the PyPi entry from authenticator, and re-creating it by scanning the QR code again. The new entry worked first time.

**Comment by NicoKiaru** [comment]
I've got the same issue (chrome, firefox, same issue), I'll try scanning the QR code again -  but that's pretty annoying.

**Comment by dehilsterlexis** [comment]
**MY FIX and wh**y. It turns out I did not finalize my authenticator app with a code. I skipped that part the first time I set it up. When using the QR code, it then asks for a finalizing step of adding the current code in the authenticator. I did not do that the first time. When I added it again, I noticed that step I had skipped. It is now working.

**Comment by dada878** [comment]
I have encountered this error as well. I tried many times, but it still shows 'Invalid TOTP code. Try again?'. However, when I tried using PyPI in incognito mode, I was successful in enabling 2 step verification.

**Comment by c9n** [comment]
> I have encountered this error as well. I tried many times, but it still shows 'Invalid TOTP code. Try again?'. However, when I tried using PyPI in incognito mode, I was successful in enabling 2 step verification.

@dada878 Wow!!! It works!! 😃 🍻

**Comment by di** [comment]
Given the similar feedback here, my best guess is that a browser extension is manipulating or invalidating the request in some way.

Given that we (the PyPI team) can't reproduce, anyone experiencing this can help by disabling/enabling extensions and retrying until they determine if one of them is causing this issue.

**Comment by dtmhdev89** [comment]
hi,
Does anyone have issue while inputting TOPT code when loggin?
![image](https://github.com/user-attachments/assets/cfd8f98f-ad55-4427-b539-119119ca0ee9)

I tried it on Firefox, Chrome browser (both normal and private mode), but it didn't work.

**Comment by di** [comment]
Hi @dtmhdev89, looks like you're having a separate issue. Please contact support@pypi.org instead.

**Comment by statmlben** [comment]
> hi, Does anyone have issue while inputting TOPT code when loggin? ![image](https://private-user-images.githubusercontent.com/39958756/359737700-cfd8f98f-ad55-4427-b539-119119ca0ee9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2MzkyODAsIm5iZiI6MTcyNDYzODk4MCwicGF0aCI6Ii8zOTk1ODc1Ni8zNTk3Mzc3MDAtY2ZkOGY5OGYtYWQ1NS00NDI3LWI1MzktMTE5MTE5Y2EwZWU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDAyMjMwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE2NDcwNDY4ZDJlNWM3OTQ2Y2U3M2U2YjU0OTBkZGUyZGQwZTQzNWQyYWNmYjc4ODczZTU1NDI1MDNkY2UyOTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Ql1X6KgKd0HSspX0RqBta1g2V1iE4aEsI2jjSWhggJk)
> 
> I tried it on Firefox, Chrome browser (both normal and private mode), but it didn't work.

I also encountered the same issue; tried Firefox, Opera, and Brave in both public and private modes on Linux and Mac; but was still unable to log in.

**Comment by dtmhdev89** [comment]
> Hi @dtmhdev89, looks like you're having a separate issue. Please contact [support@pypi.org](mailto:support@pypi.org) instead.

I had contacted the support team through above email. But hear nothing until now.

**Comment by NicoKiaru** [comment]
Here we go again, still doesn't work. I'll need a security code. Firefox, firefox private, edge. Nothing works.

**Comment by 2214372851** [comment]
I've tried everything but still can't solve it. Now my account is completely unusable.
![PixPin_2024-09-24_16-23-37](https://github.com/user-attachments/assets/be028d1d-63d6-4ffb-97cd-6d7084c35e51)

**Comment by zmievsa** [comment]
Having the same issue for the entire corporate account

**Comment by jcs090218** [comment]
Same issue here.

**Comment by konikvranik** [comment]
Hi @woodruffw , today I cannot login to pypi with the same symptoms as described in this ticket.
I use Chrome on Ubuntu. My authenticator is Bitwarden.
My locale is cs_CZ, TZ is Europe/Prague and I have time synced.

I cannot login even using neither of my backup codes.

**Comment by konikvranik** [comment]
BTW, I can log into test repo, but not to the main one.

**Comment by di** [comment]
@konikvranik Are you trying to use the same recovery/authenticator codes for both TestPyPI and PyPI? Note that they are two separate websites, so you should have separate authenticator configurations & sets of recovery codes for each.

**Comment by konikvranik** [comment]
I have just one. So it seems I overwrite main ones by test ones.
I cerated support ticket for recovery of my account.

**Comment by renn0xtek9** [comment]
Fascinating that after 1.5 years of problem they still have not pull back their 2 FA. .... When one do not want to admit its mistake, one do not want.... 

Please apply following procedure:

- Pull back immediatly 2FA (no the world did not burn before you put that in place)
- Assess the damage that you have created with this

Optional 
- Fix it 
- Re assess the damage it would create
- Find all failure modes
- Re assess again the damage it would create 
- Apply and document method for all failure mode 
- Re think if it should really be applied....
- If you cannot resist to the authoritarian IT guys that wants you to put that in place, consider 
      -  handover you resignation letter
      -  drop your pants and apply it :-S


Timezone CET 
Time.is: 0.2 seconds Delta 
Phone1: Android 14 Google TOTP 
Phone2: Android 11 Google TOTP <- account added just yesterday 
Webbrowser Google Chrome 132.0.6834.83 
OS: Ubuntu 20.04.1

```
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC=de_DE.UTF-8
LC_TIME=de_DE.UTF-8
LC_COLLATE="en_US.UTF-8"
LC_MONETARY=de_DE.UTF-8
LC_MESSAGES="en_US.UTF-8"
LC_PAPER=de_DE.UTF-8
LC_NAME=de_DE.UTF-8
LC_ADDRESS=de_DE.UTF-8
LC_TELEPHONE=de_DE.UTF-8
LC_MEASUREMENT=de_DE.UTF-8
LC_IDENTIFICATION=de_DE.UTF-8
LC_ALL=
```

I tried to:

- Remove developer option on the phone 
- Set the time completely off and then let it resync automatically (using network)


In the army they would call it sabotage... Sadly we just call it an "Open issue"....

**Comment by renn0xtek9** [comment]
Trying to so both login and totp on same android 14 phone also does not work.
Take your time to pull back 2FA...

**Comment by dkgaraujo** [comment]
Regrettably I fear I must add myself to this list. I am trying from a Windows device and from an iPhone and in neither case the 2FA code is working, even after ensuring I have the right time.

**Comment by dmx2** [comment]
This still hasn't been fixed and I've encountered this problem for over 6 months. Luckily, I stored my API key somewhere so I can still update my packages, but this is a real problem if people can't get into their accounts.

Authenticator: Authy
Browser: tried both Firefox and Chrome
OS: Arch Linux
Linux Kernel: Linux 6.14.10-arch1-1
Locale: en_US.UTF-8
Timezone: PDT

**Comment by Lutzoe** [comment]
I have the same problem. Is there any way around this?

**Comment by ykong-phd** [comment]
I had the same issue. 

-> It worked after I deleted the account on the app and re-scanned the QR code.

**Comment by iamadhee** [comment]
This one is still is an issue with my Google Authenticator. Is there a workaround for this?

**Comment by di** [comment]
We have gotten another report that this can be avoided by using Safari instead of Chrome. For anyone who finds that this is also the case, please follow my steps in https://github.com/pypi/warehouse/issues/14710#issuecomment-2271930201, otherwise there isn't much else we can do to help debug this.

**Comment by monegim** [comment]
I have exactly the same problem with the following:

- What TOTP application/authenticator are you using? Google Authenticator
- What's your browser/user agent? Google Chrome
- What's your host OS/version? Mac OS 15.6.1

Changing to Safari solved the problem.

**Comment by di** [comment]
If you're experiencing this, please follow my suggestion in https://github.com/pypi/warehouse/issues/14710#issuecomment-2271930201. Telling us that you have the same problem without giving us new information isn't helpful.

**Comment by miketheman** [comment]
Closing. If folks still have issues provisioning TOTP, please open a discrete detailed issue with as much debugging information as you can provide.
